### PR TITLE
Initial Smalltalk submission

### PR DESCRIPTION
**Pull request by avalosr**
I would like to contribute a Smalltalk chapter. I look forward to getting some feedback.

### CODE REVIEW

**Comment by aturley** [comment]
[Rendered](https://github.com/avalosr/pony-for-x/blob/chapter/smalltalk/smalltalk.asciidoc)

File: smalltalk.asciidoc [hunk 1]
```diff
@@ -0,0 +1,1617 @@
...
+
+How does Pony compare to Smalltalk? Pony is also object-oriented and functional. It is compiled into native code, not bytecode interpreted by a VM. That makes Pony fast. It is strongly and statically typed but not necessarily verbose as the compiler will infer types in most cases. It is precisely in the type system where Pony makes a novel contribution.
+
+Reference capabilities are unique to Pony. They are annotations which describe whether we can read, write or none to a variable in the current (local) scope and elsewhere (global scope). These capabilities have effects on passing and aliasing references. Thanks to _rcaps_, Pony can tell if sharing objects violates read/write laws. For example, it will prevent readers from getting an incorrect value if another thread can modify that object; it will preclude simultaneous mutators on an object. These are some of the problems that locks stop at runtime, with the penalty cost incurred by coordination. Pony detects such conflicts at compile time and won't produce an executable until they're resolved.
```
Comment (aturley) [inline smalltalk.asciidoc:16]: I think the use of the term "global scope" here might be a little confusing. It makes it sounds like reference capabilities have something to do with variable scope ("can I see this name from point X?").

I think the rest of the paragraph does a good job describing what reference capabilities do in Pony.

File: smalltalk.asciidoc [hunk 2]
```diff
@@ -0,0 +1,1617 @@
...
+Next thing to note is the `type` expression. This allows us to refer to either `Bid` or `Ask` simply as `Side`. The pony compiler will translate it to the expression on the right. We'll skip the details of the `Equatable[...]` part for now. It only means that `Bid` and `Ask` can be compared for equivalence.
+[source, pony]
+----
+primitive Bid is Equatable[Side] fun apply(): String => "Bid"
+primitive Ask is Equatable[Side] fun apply(): String => "Ask"
+type Side is (Bid | Ask)
+----
+
+=== Order
+Order objects have the same fields as the Smalltalk counterpart. Notice the use of `let` when declaring its instance variables: these fields cannot be re-assigned--unlike variables declared `val`. We need to either assign a value after declaring them or in the constructor. The compiler will complain otherwise. The naming convention indicates they are publicly accessible, which is OK in this case as they are effectively final and meant to be read.
```
Comment (aturley) [inline smalltalk.asciidoc:515]: You said: "these fields cannot be re-assigned--unlike variables declared `val`."

I believe you meant: "these fields cannot be re-assigned--unlike variables declared `var`."

File: smalltalk.asciidoc [hunk 3]
```diff
@@ -0,0 +1,1617 @@
...
+
+There is one more thing going on in the next definition. We're marking the class and constructor with the `val` reference capability. By default, classes have a `ref` capability. `ref` means _normal_ references. When you're passed a reference to an object in other languages, you can read from/write to its public fields and invoke its public methods, perhaps affecting the object's state when doing so. Here we are explicitly stating that instances of `Order` should have a `val` rcap since our intent is to use them as immutable events. We don't really have to this; we could alternatively create `Order`
```
Comment (aturley) [inline smalltalk.asciidoc:517]: You wrote: "We don’t really have to this"

I believe you meant: "We don’t really have to do this"

File: smalltalk.asciidoc [hunk 4]
```diff
@@ -0,0 +1,1617 @@
...
+  =>
+    order_id = order_id'
+    client_id = client_id'
+    sym = sym'
+    side = side'
+    price = price'
+    size = size'
+
+  fun is_bid(): Bool =>
+    side.eq(Bid)
```
Comment (aturley) [inline smalltalk.asciidoc:547]: Is there a reason you're using `eq` rather than `==` here?

File: smalltalk.asciidoc [hunk 5]
```diff
@@ -0,0 +1,1617 @@
...
+=== Records
+We declare a new type alias for our two kinds of records.
+[source, pony]
+----
+type Record is (ByPriceRecord | ByClientRecord)
+----
+
+The pattern of annotating these classes with `val` has been applied here for the same reasons mentioned earlier. The `string()` method is used to pretty-print record instances when debugging.
+
+`string()` returns a string that is currently unassigned (nothing points to it), denoted by `^`. It can be assigned only to an isolated variable, an `iso`. How is the `String ref` constructed in the method converted into a `String iso^`? The string is created and modified inside a `recover` expression. Only within `recover` is the string mutable, outside it is _frozen_, seen as having a `val` rcap.
```
Comment (aturley) [inline smalltalk.asciidoc:633]: "Only within `recover` is the string mutable, outside it is _frozen_, seen as having a `val` rcap."

Once the `String` leaves the `recover` block it has a reference capability of `iso^`, not `val`. Maybe I'm missing what you mean when you use the word "seen" here.

File: smalltalk.asciidoc [hunk 6]
```diff
@@ -0,0 +1,1617 @@
...
+
+The method itself is marked with an rcap too, `box`. This means that `string()` can be invoked by client code holding a read-only `Record`. If this sounds confusing please refer the https://tutorial.ponylang.org/capabilities/reference-capabilities.html[reference capabilities] and https://tutorial.ponylang.org/capabilities/aliasing.html[ephemeral types] sections of the Pony tutorial.
```
Comment (aturley) [inline smalltalk.asciidoc:635]: The `string()` method would have had an rcap of `box` by default, so it is worth pointing out that you are making the implicit explicit here.

**Review by aturley — COMMENTED** [review]
This looks promising. My Smalltalk isn't very good, so I'll let somebody else (like @SeanTAllen) comment on that side of things.

File: smalltalk.asciidoc (hunk 1 above)
Comment (avalosr) [reply smalltalk.asciidoc:16]: Will rework the confusing statements.

File: smalltalk.asciidoc (hunk 2 above)
Comment (avalosr) [reply smalltalk.asciidoc:515]: Correct, typo issue.

File: smalltalk.asciidoc (hunk 4 above)
Comment (avalosr) [reply smalltalk.asciidoc:547]: Should be '==' as you're pointing out. Will correct it.

File: smalltalk.asciidoc (hunk 5 above)
Comment (avalosr) [reply smalltalk.asciidoc:633]: Badly explained. Will improve.

**Comment by avalosr** [comment]
This has become out of date. Will update and re-submit in the near future.

**Pull request closed without merging**